Flexible encoders for ML-DSA

vdukhovni opened the pull request:
- Same UX as ML-KEM.  The main ASN.1 private key syntax is the one from
  Russ Housley's post on the LAMPS list, subsequently amended to tag the
  seed instead of the key (each of the three parameter sets will have a
  fixed size for the `expandedKey`):

    ML-DSA-PrivateKey ::= CHOICE {
      seed [0] IMPLICIT OCTET STRING SIZE (32),
      expandedKey OCTET STRING SIZE (2560 | 4032 | 4896)
      both SEQUENCE {
        seed OCTET STRING SIZE (32),
        expandedKey OCTET STRING SIZE (2560 | 4032 | 4896) } }

- Fixed NULL MD logic in `pkeyutl(1)` do not hardcode `EdDSA` as the only relevant algorithms.

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

mattcaswell commented:
CI failures look relevant

vdukhovni commented:
> CI failures look relevant

Thanks I neglect to commit 3 new test key files. Done.

vdukhovni commented:
Fixed test failures on Windows due to signing text files with `pkeyutl(1)` not being portable to systems with CRLF line endings.

slontis commented on apps/pkeyutl.c:
```diff
@@ -43,17 +43,15 @@ static int do_raw_keyop(int pkey_op, EVP_MD_CTX *mctx,
-static int only_rawin(const EVP_PKEY *pkey)
-{
-    return is_EdDSA(pkey);
+    deftype = EVP_PKEY_get_default_digest_name(pkey, defname, sizeof(defname));
```
funnily enough - I was going to add  support for this today..

slontis commented on crypto/ml_dsa/ml_dsa_encoders.c:
```diff
@@ -768,7 +766,13 @@ int ossl_ml_dsa_sk_decode(ML_DSA_KEY *key, const uint8_t *in, size_t in_len)
 
-    if (key->priv_encoding != NULL || key->pub_encoding != NULL)
+    /* When loading from an explicit key, drop the seed. */
+    OPENSSL_free(key->seed);
```
Why cant there be both?

slontis replied:
raised in the issue - so closing for now..

vdukhovni replied:
Because if we did not load the seed, we have no idea that it corresponds to the key we did use.

slontis commented on crypto/ml_dsa/ml_dsa_encoders.c:
```diff
@@ -768,7 +766,13 @@ int ossl_ml_dsa_sk_decode(ML_DSA_KEY *key, const uint8_t *in, size_t in_len)
+    key->seed = NULL;
+
+    /* Allow the key encoding to be already set to the provided pointer */
+    if ((key->priv_encoding != NULL && key->priv_encoding != in)
```
Why does key->priv_encoding == in happen (has it already been processed in this case?)

vdukhovni replied:
This happens when the key has been decoded, but not yet loaded or imported. A separate pointer could be used, and "moved" (set one clear the other) so that the "prekey" is more explicit. I am not opposed if you think that's warranted, already commented on this...

slontis replied:
It not particularly easy to follow either way :)

vdukhovni replied:
If you have a specific plan and time to execute it, I'm not opposed to modest refactoring. What do you have in mind?

slontis commented on crypto/ml_dsa/ml_dsa_key.c:
```diff
@@ -92,7 +148,14 @@ void ossl_ml_dsa_key_free(ML_DSA_KEY *key)
     OPENSSL_free(key->pub_encoding);
-    OPENSSL_free(key->priv_encoding);
+    if (key->priv_encoding != NULL) {
+        OPENSSL_cleanse(key->priv_encoding, key->params->sk_len);
```
Use OPENSSL_clear_free() instead of this pattern
same below..

slontis commented on crypto/ml_dsa/ml_dsa_key.c:
```diff
@@ -139,13 +209,15 @@ ML_DSA_KEY *ossl_ml_dsa_key_dup(const ML_DSA_KEY *src, int selection)
+                        goto err;
                 }
+                if (src->seed != NULL
+                    && (ret->seed = OPENSSL_memdup(src->priv_encoding,
```
src->priv_encoding? (Doesnt look like it is hitting this path when testing)

vdukhovni replied:
Good catch. Dup of keys that already have key material is rather rare at the EVP layer. So this slipped through.
Thanks!

slontis commented on crypto/ml_dsa/ml_dsa_key.c:
```diff
@@ -379,14 +400,11 @@ int ossl_ml_dsa_key_pairwise_check(const ML_DSA_KEY *key)
  *
- * @param seed The input seed
- * @param seed_len The size of entropy (Must be 32 bytes)
  * @param out The generated key (which contains params on input)
```
which includes the seed

vdukhovni replied:
The seed is no longer a separate argument, it set into the "prekey" by a separate call. What specific language would you like to see here?

slontis replied:
```suggestion
 * @param out The generated key (which contains params on input, including the seed)
```

slontis commented on crypto/ml_dsa/ml_dsa_key.c:
```diff
@@ -421,50 +439,51 @@ static int keygen_internal(const uint8_t *seed, size_t seed_len,
+
 err:
+    if (out->seed != NULL && !out->retain_seed) {
+        OPENSSL_cleanse(out->seed, ML_DSA_SEED_BYTES);
```
OPENSSL_clear_free()

vdukhovni replied:
Sure, coming up. In GSSAPI, you free pointers by reference, and they're also set to NULL for you, here I still have to do that explicitly.

slontis commented on doc/man7/EVP_PKEY-ML-DSA.pod:
```diff
@@ -1,5 +1,7 @@
 =pod
 
+=encoding utf-8
```
Why is this necessary? (it is not in any other pod file)

vdukhovni replied:
Actually it is.

slontis replied:
My concern was if this is something new then it might not output correctly  of the web pages.

slontis commented on doc/man7/EVP_PKEY-ML-DSA.pod:
```diff
@@ -69,6 +79,127 @@ respective key type of B<ML-DSA-44>, B<ML-DSA-65> or B<ML-DSA-87>.
+=item C<priv-only>:
+
+This format represents B<PKCS#8> objects in which only the FIPS 204
+decapsulation key B<sk> is present in the above sequence.
```
decapsulation?

vdukhovni replied:
Oops, right that's for ML-KEM not ML-DSA, fix coming up...

slontis commented on doc/man7/EVP_PKEY-ML-DSA.pod:
```diff
@@ -69,6 +79,127 @@ respective key type of B<ML-DSA-44>, B<ML-DSA-65> or B<ML-DSA-87>.
+If the C<priv-only> format is not included in the list, this format will not be
+recognised on input.
+
+=item C<oqs-kpair>:
```
why kpair?

vdukhovni replied:
Well it is a key pair, (private + public). Bikeshedding the name is guess OK, if you feel there's an obviously better name.

slontis replied:
pair or keypair

slontis commented on doc/man7/EVP_PKEY-ML-DSA.pod:
```diff
@@ -69,6 +79,127 @@ respective key type of B<ML-DSA-44>, B<ML-DSA-65> or B<ML-DSA-87>.
+=item C<bare-seed>:
+
+This format represents B<PKCS#8> objects in which the private key contains
+the 32-byte FIPS 204 seed B<ξ> without any ASN.1 encapsulation.
```
It should be clear without using B<ξ>

vdukhovni replied:
Is there something wrong with ξ?

slontis replied:
thats why you did utf8 on this file - so its related to the other question

vdukhovni replied:
Yes, the world is UTF-8 now. When it makes sense I am not shackled to US-ASCII.
So we can also talk about `external μ`, ... without being reticent about matching the actual terms in FIPS 203/204/...

slontis commented on providers/implementations/encode_decode/ml_dsa_codecs.c:
```diff
@@ -0,0 +1,575 @@
+      0x01, 0x65, 0x03, 0x04, 0x03, 0x13, 0x03, 0x82, 0x0a, 0x21, 0x00, }
+};
+static const ML_DSA_PKCS8_FMT ml_dsa_87_p8fmt[NUM_PKCS8_FORMATS] = {
+    { "seed-priv", 0x134a, 0, 0x30821346, 0x8020, 6, 0x20, 0x04821320, 0x2a, 0x1320, 0,      0,     },
```
Not really sure why you chose this approach.. The ANS1 d2i and i2d would have been sufficient,,,

e.g. (not exactly what you have done here - but it illustrates the point).

````
typedef struct ASN1_PQ_PRIV_KEY_st {
    ASN1_OCTET_STRING *seed;
    ASN1_OCTET_STRING *priv;
} ASN1_PQ_PRIV_KEY;

typedef struct ASN1_PQ_PRIV_KEY_IN_st {
    int type;
    union {
        ASN1_PQ_PRIV_KEY *def;
        ASN1_OCTET_STRING *oqs;
    } value;
} ASN1_PQ_PRIV_KEY_IN;

/* ML_DSA Private key output format */
ASN1_SEQUENCE(ASN1_PQ_PRIV_KEY) = {
    ASN1_OPT(ASN1_PQ_PRIV_KEY, seed, ASN1_OCTET_STRING),
    ASN1_IMP_OPT(ASN1_PQ_PRIV_KEY, priv, ASN1_OCTET_STRING, 1)
} static_ASN1_SEQUENCE_END(ASN1_PQ_PRIV_KEY)

#define d2i_ASN1_PQ_PRIV_KEY(a, out) \
    ASN1_item_i2d((const ASN1_VALUE *)a, out, ASN1_ITEM_rptr(ASN1_PQ_PRIV_KEY))
IMPLEMENT_STATIC_ASN1_ALLOC_FUNCTIONS(ASN1_PQ_PRIV_KEY)

/* ML-DSA Private key input format allows a choice of formats */
#define ASN1_PQ_PRIV_KEY_IN_TYPE_DEFAULT 0
#define ASN1_PQ_PRIV_KEY_IN_TYPE_OQS     1
ASN1_CHOICE(ASN1_PQ_PRIV_KEY_IN) = {
    ASN1_SIMPLE(ASN1_PQ_PRIV_KEY_IN, value.def, ASN1_PQ_PRIV_KEY),
    ASN1_OPT(ASN1_PQ_PRIV_KEY_IN, value.oqs, ASN1_OCTET_STRING),
} static_ASN1_CHOICE_END(ASN1_PQ_PRIV_KEY_IN)

#define d2i_ASN1_PQ_PRIV_KEY_IN(a, in, len) \
    (ASN1_PQ_PRIV_KEY_IN *)ASN1_item_d2i((ASN1_VALUE **)a, in, len, ASN1_ITEM_rptr(ASN1_PQ_PRIV_KEY_IN))
IMPLEMENT_STATIC_ASN1_ALLOC_FUNCTIONS(ASN1_PQ_PRIV_KEY_IN)

/*
 * Only save out the non OQS format
 * This just conditionally writes the seed and priv components and then does
 * an i2d.
 */
int ossl_i2d_ml_dsa_priv(const ML_DSA_KEY *key, uint8_t **der)
{
    int len = 0;
    ASN1_PQ_PRIV_KEY *obj;
    const uint8_t *seed, *priv;

    priv = ossl_ml_dsa_key_get_priv(key);
    if (priv == NULL)
        return 0;
    obj = ASN1_PQ_PRIV_KEY_new();
    if (obj == NULL)
        return 0;

    obj->priv = ASN1_OCTET_STRING_new();
    if (obj->priv == NULL
            || !ASN1_OCTET_STRING_set(obj->priv, priv,
                                      ossl_ml_dsa_key_get_priv_len(key)))
        goto err;

    seed = ossl_ml_dsa_key_get_seed(key);
    if (seed != NULL) {
        if ((obj->seed = ASN1_OCTET_STRING_new()) == NULL
                || !ASN1_OCTET_STRING_set(obj->seed, seed,
                                          ossl_ml_dsa_key_get_seed_len(key)))
            goto err;
    }
    len = i2d_ASN1_PQ_PRIV_KEY(obj, &der);
err:
    ASN1_PQ_PRIV_KEY_free(obj);
    return len;
}

ML_DSA_KEY *ossl_d2i_ml_dsa_priv(const uint8_t *in, int len)
{
    int ok = 0;
    ML_DSA_KEY *key = NULL;
    const uint8_t *pder = in;
    ASN1_PQ_PRIV_KEY_IN *data = d2i_ASN1_PQ_PRIV_KEY_IN(NULL, &pder, len);

    if (in != NULL) {
        if (data->type == ASN1_PQ_PRIV_KEY_IN_TYPE_DEFAULT) {
            /* default */
            ASN1_PQ_PRIV_KEY *def = data->value.def;
            ASN1_OCTET_STRING *seed = def->seed;
            ASN1_OCTET_STRING *priv = def->priv;
````

vdukhovni replied:
The legacy OQS format is distinguished from the private key only by length, where we have to know to expect an extra |pubkey| bytes on the end of the private key. I don't believe our ASN.1 templates are rich enough to express this (they're not an ASN.1 2021 compiler). Also, since the IETF LAMPS group is still debating, we may end up with formats outside ASN.1, hard to day, the code supports "bare" keys, which are still being advocated for rather odd reasons.

vdukhovni replied:
Anyway, we have working code. It is flexible, we can likely move on from the internals of the parser to more meaningful decisions, e.g.:
What might make sense is to define the output structure of the parse explicitly as a component of a parsed but not yet imported/loaded key. You have a tentative definition above.
There could be NULLable fields in the key that contain separate references to the embryonic "prekey" elements (used on export of keys that have no real key material).
At the moment, the fields used for long-term seed retention and the encoded private key are overloaded to hold the embryonic key, which minimised changes to the key format, but is perhaps a bit forced.
If you want to argue for an explicit pair of separate pointers (preseed, prekey) in the key object (I don't think we need to allocate a pair to hold that, these can be directly embedded), with everything else implemented it is probably not difficult to do that. Do you have cycles to refactor the key structure in that direction as a follow-on PR?
Perhaps during the pre-freeze window, if the decision to merge the branch is made, but this sort of cleanup is fine.

slontis replied:
yep so you read it into the object and then look at the length of the field read... no brainer..

slontis commented on providers/implementations/encode_decode/ml_dsa_codecs.h:
```diff
@@ -0,0 +1,103 @@
+ * The p8_shift is 0 when the top-level tag+length occupy four bytes, 2 when
+ * they occupy two by†es, and 4 when no tag is used at all.
+ */
+typedef struct {
```
later down the track we can share this object

vdukhovni replied:
Yes, when merged the `PKCS#8` format type and preference vectors are not algorithm specific and can be shared.

slontis commented on providers/implementations/keymgmt/ml_dsa_kmgmt.c:
```diff
@@ -258,52 +351,76 @@ static int ml_dsa_export(void *keydata, int selection,
-     * Note that the private key always contains the public key elements so we
-     * just save the one blob and return.
+     * Note that the public key can be recovered from the private key, so we
+     * just export one or the other.  If the seed is present, both the seed and
```
I dont know if this is the correct decision....

vdukhovni replied:
I believe I retained the original behaviour (one blob or the other). We could export both, which is redundant for our providers, which always recomputed the public key, but arguably useful for exporting to others. I have no strong views on whether to include the public component in `export`. It is easier to add later, than remove, so starting with just one leaves more options open perhaps, otherwise toss a coin.

slontis replied:
I have raised an issue with a blocker on merge to master tag.

slontis replied:
not a blocker here..

slontis commented on include/openssl/core_dispatch.h:
```diff
@@ -19,6 +19,13 @@
+ * functions of various signatures must occupy a common slot in an array of
+ * structures.
+ */
+typedef void (*OSSL_FUNC)(void);
```
Why did you add this just for the load dispatch table entry?

vdukhovni replied:
It is used broadly in the ML-KEM case, and already slated to go in.  This just does the same thing in the ML-DSA branch, which will just use the same construct when both are merged.

slontis commented:
Approved with or without nit fixes.

slontis commented:
reapproved

t-j-h reviewed (approved):
That went in cleanly!

vdukhovni commented:
Merged Shane's prior PR to the branch, and rebased the feature branch on master.  Then rebased this PR on the updated feature branch. This is now self-contained.
Added a small fixup with the current snapshot of the ASN.1 in the doc.

vdukhovni commented:
Merged. Many thanks for the reviews.
